### PR TITLE
Add permanently denied option to listener

### PR DESCRIPTION
**Pull request by anothem**
As discussed in https://github.com/Karumi/Dexter/issues/4

### CODE REVIEW

**Comment by pedrovgs** [comment]
Thanks for your PR @anothem!!! :+1: we will wait for @Serchinastico's approval before to merge the branch.

**Comment by Serchinastico** [comment]
Nice solution!

I'd change one thing though, it's something we discussed in another PR. Given it's clear we need to change the listener to add these new cool features I'd do it in a different fashion. I'd rather send back a ``PermissionRequest`` object in the current three ``PermissionListener`` methods instead of a plain String with the name (I'd like to think about creating different subclasses for each method, but that's something for future development). In this way we can add more information to the listeners while keeping backwards compatibility (from the moment we implement it :P). In the other hand, adding more methods to the listener will force subclasses to implement things they might not need at all and will surely break their code with every library upgrade.

Do you mind implementing the callback in this way? To sum up:

* Create a new ``PermissionRequest`` class field holding the name of the requested permission and a boolean with your perma-denied flag.
* Change ``PermissionListener`` signature to return the new object in every method instead of the permission name.

Thanks for the PR @anothem!

**Comment by anothem** [comment]
If we change the parameter type on the listener methods, it would break backwards compatibility anyway, and would require a major version bump. But I guess you're trying to avoid having to do this again in the future, right? 

If we do a wrapper object, perhaps it makes more sense to do one for each scenario? e.g. `PermissionDeniedResponse` and `PermissionGrantedResponse`. 
If we just use one generic wrapper class, I'm concerned about it having fields that wouldn't make sense in all the contexts where it's used (e.g. `boolean neverAskAgainSelected` makes sense for when it's denied, but not for when it's accepted).

**Comment by Serchinastico** [comment]
Yep,  we are going to break backwards compatibility either way so let's try to do it better this time :)

I have the same concerns and that's what I thought as well: seems like a better solution to return 3 different response objects, I just wanted to set aside some time to think if they should share some basic data and, if they do, in which way. 

Personally, I'd create 3 different response objects containing an instance of a PermissionRequest class (holding request information like the name and such to keep track of the source of the response). Every Response class would also contain response-specifics like the perma-denied flag or whatever feature we can think of. 

I think in this way we can safely continue extending the library.

What do you think? @pedrovgs & @anothem

**Comment by Serchinastico** [comment]
The ``PermissionRequest`` object is not really needed for this PR, I'm just thinking ahead FYI :P

**Comment by pedrovgs** [comment]
@Serchinastico your approach seems correct. @anothem can we implement the minimum changes needed to match with @Serchinastico's approach to merge this PR and publish a new library version?

**Comment by anothem** [comment]
Yes, I'll have a look and update :)

**Comment by anothem** [comment]
Alright, I gave it a shot based on what we talked about. Let me know what you think :) 
@Serchinastico there seemed little point in making an extra response object for permanently-denied, since it was being returned in the same method, so I did a single response one with a boolean flag. Hope I didn't misunderstood what you intended.

File: dexter/src/main/java/com/karumi/dexter/listener/PermissionDeniedResponse.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
+package com.karumi.dexter.listener;
```
Comment (Serchinastico) [inline dexter/src/main/java/com/karumi/dexter/listener/PermissionDeniedResponse.java]: Missing copyright

Comment (Serchinastico) [inline dexter/src/main/java/com/karumi/dexter/listener/PermissionDeniedResponse.java]: Same for the other new classes

File: dexter/src/main/java/com/karumi/dexter/listener/PermissionDeniedResponse.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+  public PermissionDeniedResponse(@NonNull PermissionRequest requestedPermission, boolean permanentlyDenied) {
+    this.requestedPermission = requestedPermission;
+    this.permanentlyDenied = permanentlyDenied;
+  }
+
+  /**
+   * Builds a new instance of PermissionDeniedResponse from a given permission string
+   * and a permanently-denied boolean flag
+   *
+   * @param permission
```
Comment (Serchinastico) [inline dexter/src/main/java/com/karumi/dexter/listener/PermissionDeniedResponse.java]: Remove these useless param & return comments

**Comment by Serchinastico** [comment]
Besides the very minor issues the PR looks great to me! :+1:

**Comment by Serchinastico** [comment]
Thanks @anothem! Great PR & discussion!

**Pull request merged**